### PR TITLE
Start of Integer flags implementation

### PR DESCRIPTION
**Pull request by behunin**
This is the start of implementing shader_ir flags.
I have only tested this on Breath of the Wild.

### CODE REVIEW

File: src/video_core/shader/shader_ir.h [hunk 1]
```diff
@@ -346,6 +346,9 @@ class ShaderIR final {
     /// Access a bindless image sampler.
     Image& GetBindlessImage(Tegra::Shader::Register reg, Tegra::Shader::ImageType type);
 
+    /// Recursive Iteration over the OperationNode operands
+    void Iterop(NodeBlock& nb, Node var);
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.h:350]: This should likely have a more informative name, Like `RecursiveOpIterate` or something.

File: src/video_core/shader/shader_ir.cpp [hunk 2]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
     if (!sets_cc) {
         return;
     }
-    Node zerop = Operation(OperationCode::LogicalIEqual, std::move(value), Immediate(0));
-    SetInternalFlag(bb, InternalFlag::Zero, std::move(zerop));
-    LOG_WARNING(HW_GPU, "Condition codes implementation is incomplete");
+    switch (value->index()) {
+    case 0:
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:391]: Magic values

File: src/video_core/shader/shader_ir.cpp [hunk 3]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
...
+        Iterop(bb, value);
+        break;
+    case 2:
+        if (const auto gpr = std::get_if<GprNode>(value.get())) {
+            LOG_WARNING(HW_GPU, "GprNode: index={}", gpr->GetIndex());
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:396]: This should be a debug log

File: src/video_core/shader/shader_ir.cpp [hunk 4]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
...
+    default:
+        Node zerop = Operation(OperationCode::LogicalIEqual, std::move(value), Immediate(0));
+        SetInternalFlag(bb, InternalFlag::Zero, std::move(zerop));
+        LOG_WARNING(HW_GPU, "Node Type: {}", value->index());
+        break;
+    }
+}
+
+void ShaderIR::Iterop(NodeBlock& nb, Node var) {
+    if (const auto op = std::get_if<OperationNode>(var.get())) {
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:412]: We generally use guard clauses to reduce nesting where possible:

```suggestion
    const auto* op = std::get_if<OperationNode>(var.get());
    if (op == nullptr) {
        return;
    }
    
    if (op->GetOperandsCount() == 0) {
        return;
    }
    
    // rest of the code here
```

File: src/video_core/shader/shader_ir.cpp [hunk 5]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
...
+        if (op->GetOperandsCount() > 0) {
+            for (auto& opss : op->GetOperands()) {
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:414]:
```suggestion
            for (auto& operand : op->GetOperands()) {
```

File: src/video_core/shader/shader_ir.cpp [hunk 6]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
...
+                switch (opss->index()) {
+                case 0:
+                    return Iterop(nb, opss);
+                case 2:
+                    if (const auto gpr = std::get_if<GprNode>(opss.get())) {
+                        LOG_WARNING(HW_GPU, "Child GprNode: index={}", gpr->GetIndex());
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:420]: Debug log

File: src/video_core/shader/shader_ir.cpp [hunk 7]
```diff
@@ -387,9 +387,49 @@ void ShaderIR::SetInternalFlagsFromInteger(NodeBlock& bb, Node value, bool sets_
...
+
+void ShaderIR::Iterop(NodeBlock& nb, Node var) {
+    if (const auto op = std::get_if<OperationNode>(var.get())) {
+        if (op->GetOperandsCount() > 0) {
+            for (auto& opss : op->GetOperands()) {
+                switch (opss->index()) {
+                case 0:
+                    return Iterop(nb, opss);
+                case 2:
+                    if (const auto gpr = std::get_if<GprNode>(opss.get())) {
```
Comment (lioncash) [inline src/video_core/shader/shader_ir.cpp:419]:
```suggestion
                    if (const auto* gpr = std::get_if<GprNode>(opss.get())) {
```

File: src/video_core/shader/node.h [hunk 8]
```diff
@@ -464,6 +464,10 @@ class OperationNode final : public AmendNode {
         return operands.size();
     }
 
+    NodeBlock GetOperands() const {
```
Comment (lioncash) [inline src/video_core/shader/node.h:467]:
```suggestion
    const NodeBlock& GetOperands() const {
        return operands;
    }
```

So usages of this function don't churn std::vector allocations.

File: src/video_core/shader/shader_ir.h (hunk 1 above)
Comment (behunin) [reply src/video_core/shader/shader_ir.h:350]: I like that suggestion, here is a few more: OperandsIteration, RecurseOperands, SearchOperands.
Whatever seems clever to the team.

File: src/video_core/shader/shader_ir.cpp (hunk 2 above)
Comment (behunin) [reply src/video_core/shader/shader_ir.cpp:391]: This is the std::variant<Types...>::index, so would a comment above the switch statement be sufficient enough to clarify?

Comment (behunin) [reply src/video_core/shader/shader_ir.cpp:391]: // value->index() is the index of the std::variant, tells us which Node type we have here.

What ya think?

File: src/video_core/shader/shader_ir.h (hunk 1 above)
Comment (behunin) [reply src/video_core/shader/shader_ir.h:350]: /// Recursive Iteration over the OperationNode operands, searching for a GprNode.

I am favoring SearchOperands now. Cool?

File: src/video_core/shader/shader_ir.cpp (hunk 2 above)
Comment (lioncash) [reply src/video_core/shader/shader_ir.cpp:391]: This is sort of self-explanatory without really elaborating on the node type. It's likely more informative to add a comment by the number like:

```cpp
case 0: // Operation node
    ...
````

**Comment by behunin** [comment]
I don't know why CI didn't build successfully. Can someone please trigger the build again?

**Comment by v1993** [comment]
@behunin I think you can rebase on top of new commits in master and force-push to trigger a CI rebuild.

CI build have completed, but [tests have failed](https://dev.azure.com/yuzu-emu/yuzu/_build/results?buildId=7957&view=logs&j=b457fee1-24b5-5b05-276f-95f55538c81f&t=8274a14c-2774-5fa1-24fc-2bf7929518a4&l=3485) for some reason.

**Comment by behunin** [comment]
@v1993 Thank you! I see now it's just tests that failed, I am not really worried about that and am guessing the team feels the same.

**Comment by behunin** [comment]
@bunnei @ReinUsesLisp @lioncash
Good to go er nah?

**Review by bunnei — APPROVED** [review]

**Review by adityaruplaha — COMMENTED** [review]
I know this has been merged already, and it's a really minor thing but...

**Pull request merged**